lib/data/repository/dio_marvell_repository.dart

```diff
@@ -1,84 +1,103 @@
+    _repository.remoteDataSource = remoteDataSource;
     return _repository;
   }
 
-  DioMarvellRepository._internal();
-
   @override
-  Future<List<Character>> getCharacters([offset = 0]) async {
+  Future<Character> getCharacterDetail(String characterId) async {
     try {
-      var httpResponse = await marvellClient.getCharacters(fakeTimeStamp, publicKey, hash, offset);
+      var httpResponse = await remoteDataSource.getCharacterDetail(characterId);
       if (httpResponse.response.statusCode == 200) {
         var data = httpResponse.data.data;
-        if (data != null) {
-          List<Character>? tempCharacters = data.results?.map((item) => item.toCharacter("standard_medium")).toList();
-          List<Character> characters = tempCharacters ?? List.empty();
-          return characters;
-        } else {
-          return Future.error("No data error");
+        try {
+          Character? tempCharacter = data?.results?.map((item) => item.fromApiToCharacter("standard_xlarge")).single;
+          Character character = tempCharacter as Character;
+          return character;
+        } on Exception {
+          throw DataParsingException();
         }
       } else {
-        return Future.error("network error");
+        throw DataRetrieveException();
       }
     } catch (error) {
-      return Future.error("$error");
+      if ((error is DataRetrieveException) || (error is DataParsingException)) {
+        rethrow;
+      } else {
+        throw NoConnectionException();
+      }
     }
   }
 
   @override
-  Future<Character> getCharacterDetail(String characterId) async {
+  Future<List<Character>> getCharacters([int offset = 0]) async {
     try {
-      var httpResponse = await marvellClient.getCharacterDetail(characterId, fakeTimeStamp, publicKey, hash);
+      var httpResponse = await remoteDataSource.getCharacters(offset);
       if (httpResponse.response.statusCode == 200) {
         var data = httpResponse.data.data;
-        if (data != null) {
-          Character? tempCharacter = data.results?.map((item) => item.toCharacter("standard_xlarge")).single;
-          Character character = tempCharacter as Character;
-          return character;
-        } else {
-          return Future.error("No data error");
+        try {
+          List<Character>? tempCharacters =
+          data?.results?.map((item) => item.fromApiToCharacter("standard_medium")).toList();
+          List<Character> characters = tempCharacters ?? List.empty();
+          localDataSource.characterDao.deleteOldData();
+          localDataSource.characterDao.insertData(characters);
+          return characters;
+        } on Exception {
+          throw DataParsingException();
         }
       } else {
-        return Future.error("network error");
+        throw DataRetrieveException();
       }
     } catch (error) {
-      return Future.error("$error");
+      if ((error is DataRetrieveException) || (error is DataParsingException)) {
```

Comment:
Logic with recatching and rethrowing looks confusing, we need to refactor it to simplify or remove it. It looks like you are not using these different types in your BLoC's

Reply:
done